pkg/bgpv1/manager/reconcilerv2/policies.go

```diff
@@ -197,6 +197,106 @@ func CreatePolicy(name string, peerAddr netip.Addr, v4Prefixes, v6Prefixes types
 	return policy, nil
 }
 
+// MergeRoutePolicies evaluates two instances of RoutePolicy{} and returns a single RoutePolicy{} representing
+// the merger of the two.  The merge operation focuses on each policy's Statements.  Statements define one or more
+// Actions, and these actions may include setting BGP Communities, Local Preference, and others.  Statements are
+// keyed by their Conditions, which define the BGP Neighbor, AF, and Prefixes it applies to.
+//
+// The merge operation evaluates Actions across only those statements with the same key. For these Statements,
+// the merge takes the union of all BGP Communities set.  When differing Local Preference values are set, the
+// higher value is selected.
+func MergeRoutePolicies(policyA *types.RoutePolicy, policyB *types.RoutePolicy) (*types.RoutePolicy, error) {
+	if policyA == nil || policyB == nil {
+		return nil, fmt.Errorf("route policy is nil")
+	}
+	if policyA.Name != policyB.Name {
+		return nil, fmt.Errorf("route policy names do not match")
+	}
+	if policyA.Type != policyB.Type {
+		return nil, fmt.Errorf("route policy types do not match")
+	}
+
+	mergedPolicy := &types.RoutePolicy{
+		Name:       policyA.Name,
+		Type:       policyA.Type,
+		Statements: []*types.RoutePolicyStatement{},
+	}
+
+	// Maps a string key representing a unique RoutePolicyConditions{} (where unique is defined by the values of all
+	// attributes) to a RoutePolicyStatement{}.  When multiple instances of the same key representing
+	// RoutePolicyConditions{} are observed, the existing RoutePolicyStatement{} will be updated to reflect the union
+	// of all instances.
+	conditionsToMergedRoutePolicyStatement := map[string]*types.RoutePolicyStatement{}
+
+	// Extract the first policy's statements and communities
+	for _, statement := range policyA.Statements {
+		key := statement.Conditions.String()
+		if _, found := conditionsToMergedRoutePolicyStatement[key]; !found {
+			conditionsToMergedRoutePolicyStatement[key] = &types.RoutePolicyStatement{
+				Actions:    statement.Actions,
+				Conditions: statement.Conditions,
+			}
+		} else {
+			if len(statement.Actions.AddCommunities) > 0 {
+				conditionsToMergedRoutePolicyStatement[key].Actions.AddCommunities = append(
+					conditionsToMergedRoutePolicyStatement[key].Actions.AddCommunities, statement.Actions.AddCommunities...)
+			}
+			if len(statement.Actions.AddLargeCommunities) > 0 {
+				conditionsToMergedRoutePolicyStatement[key].Actions.AddLargeCommunities = append(
+					conditionsToMergedRoutePolicyStatement[key].Actions.AddLargeCommunities, statement.Actions.AddLargeCommunities...)
+			}
+			// RFC 4271 states "The higher degree of preference MUST be preferred."
+			if statement.Actions.SetLocalPreference != nil &&
+				*statement.Actions.SetLocalPreference > *conditionsToMergedRoutePolicyStatement[key].Actions.SetLocalPreference {
```

Comment:
`*conditionsToMergedRoutePolicyStatement[key].Actions.SetLocalPreference` can be nil de-reference as well, right?

Reply:
Is this what you are thinking about?
```
var nilPointerInt64 *int64

policyA: &types.RoutePolicy{
Name: "policy",
Statements: []*types.RoutePolicyStatement{
	{
		Conditions: types.RoutePolicyConditions{},
		Actions: types.RoutePolicyActions{
			RouteAction:        types.RoutePolicyActionAccept,
			SetLocalPreference: nilPointerInt64,
		},
	},
},
```

Reply:
I think both are same, this is the test which would cause panic

```
		{
			name: "nil dereferenced statement.Action.SetLocalPref",
			policyA: &types.RoutePolicy{
				Name: "policy",
				Statements: []*types.RoutePolicyStatement{
					{
						Conditions: types.RoutePolicyConditions{},
						Actions: types.RoutePolicyActions{
							RouteAction:        types.RoutePolicyActionAccept,
							SetLocalPreference: nil, // can be nil or nilPointerInt64
						},
					},
				},
			},
			policyB: &types.RoutePolicy{
				Name: "policy",
				Statements: []*types.RoutePolicyStatement{
					{
						Conditions: types.RoutePolicyConditions{},
						Actions: types.RoutePolicyActions{
							RouteAction:        types.RoutePolicyActionAccept,
							SetLocalPreference: ptr.To[int64](100), // In policy B, have some local pref value
						},
					},
				},
			},
			errorExpected: false,
			expected: &types.RoutePolicy{
				Name: "policy",
				Statements: []*types.RoutePolicyStatement{
					{
						Conditions: types.RoutePolicyConditions{},
						Actions: types.RoutePolicyActions{
							RouteAction:        types.RoutePolicyActionAccept,
							SetLocalPreference: ptr.To[int64](100),
						},
					},
				},
			},
		},
```

Reply:
TY!  I repro'd the panic with that one:
```
--- FAIL: Test_MergeRoutePolicies (0.00s)
    --- FAIL: Test_MergeRoutePolicies/nil_dereferenced_statement.Action.SetLocalPref (0.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x2dea3c2]
```